File: Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        set<char>a;
        int end=0,start=0,max=0;
        while(start<s.size())
        {
            if(a.find(s[start])==a.end())
            {a.insert(s[start]);
            if(max<start-end+1)
            max=start-end+1;
            start++;
            }
            else{
                a.erase(s[end]);
                end++;
            }
        }
        return max;
    }

};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s.isEmpty())return 0;
        HashSet<Character>st = new HashSet<>();
        int ans=0,i=0,j=0;
        while(j<s.length()){
            if(st.isEmpty()){
                st.add(s.charAt(j));
                ans=Math.max(ans,1);
                j++;
            }
            else{
                int iniLen = st.size();
                st.add(s.charAt(j));
                if(st.size()==iniLen){
                    while(i<s.length()&&s.charAt(i)!=s.charAt(j)){
                        st.remove(s.charAt(i));
                        i++;
                    }
                    st.add(s.charAt(j));
                    i++;
                    j++;
                }
                else {
                    ans=Math.max(ans,j-i+1);
                    j++;
                }
            }

        }
        return ans;
    }
}